lib/daemon.py

```diff
@@ -18,16 +19,26 @@ def main() -> int:
     # Make sure the ua-daemon logger does not generate double logging
     # by propagating to the root logger
     LOG.propagate = False
-    # The root logger should only log errors to the daemon log file
-    setup_logging(
-        logging.CRITICAL, logging.ERROR, log_file=cfg.daemon_log_file
-    )
+    # TODO: keep an eye on logs as we add features to make sure they don't
+    # fill up too much
+    setup_logging(logging.INFO, logging.DEBUG, log_file=cfg.daemon_log_file)
 
     LOG.debug("daemon started")
 
+    daemon.on_start(cfg)
+
+    threads = daemon.start_background_threads(cfg)
+    LOG.debug("started {} threads".format(len(threads)))
+
+    # TODO: should this go earlier?
     notify("READY=1")
```

Comment:
Thanks! I was hoping to get your opinion on this to make sure It seems right.

`main_thread` currently does nothing, but it will eventually be any work that will happen on the main thread of the daemon. It will be blocking.
In other words, getting to this point means that we've finished all start up tasks. Anything in `main_thread` will be ongoing background tasks.

My other concern is that `on_start` usually will take no time, but on rare occasion might take a while (a few minutes). Is there any reason that taking a couple minutes to call `notify` could cause problems?

Reply:
My concern is in the opposite direction: I am wondering if the service is up and running by the time sd_notify is called.

For example, let's say this was a web server. I'd only call notify once the service was already listening to new connections (instead of right after the function that binds to a port is called), if possible. In case a service depends on this one, we should guarantee the service is up and running by the time notify is called.

I do not see major issues on longer starting times now and then. @sergiodj, WDYT?

Reply:
Right that makes sense.

I think that because, at this stage, the daemon is not providing a service for other software to interact with directly, the timing of `notify` is okay as it is.

In the future we may implement a dbus API or similar, at which point we will need to coordinate that the API is listening before `notify`.

Even in the future, `main_thread` should only be ua-internal tasks that are irrelevant to other programs. So I think it is okay to call `notify` before `main_thread`. I'll add this to the docstring for `main_thread` so that we remember this in the future.

Reply:
Hey, I'm joining the party a bit late, sorry about that.

So, if I understand correctly `main_thread` is almost a no-op right now, and in the future it will only take care of internal ua tasks, right?  Just by reading its name, I actually thought it implemented the main loop of the program or some such (in which case I'd agree with Athos that `READY=1` should only be issued after `main_thread` is finished).  What would you think about renaming it to something less "important"?

Also, interesting that you mention your plans to implement a DBus API.  systemd offers a `Type=dbus` which may come in handy for you.

Reply:
@sergiodj All good, thank you! :)

You're right about `main_thread`. I only named it based on what thread it would be running on but do realize now that it also implies high significance. I did add a docstring to the function indicating that it is not providing any services. But we can rename it as well. Maybe something like `internal_main_thread_tasks`? or just `internal_tasks`?

Ah yes I saw the `Type=dbus` in the docs, but a DBus API is still in the idea stage and would happen 6 months from now at the earliest. So I didn't think it was worth doing the work now to be a proper dbus service. Do you foresee problems of transitioning from type `notify` to `dbus` in the future?

Reply:
> You're right about `main_thread`. I only named it based on what thread it would be running on but do realize now that it also implies high significance. I did add a docstring to the function indicating that it is not providing any services. But we can rename it as well. Maybe something like `internal_main_thread_tasks`? or just `internal_tasks`?

Gotcha.  Based on your description earlier, I'd say that `internal_tasks` sounds like a good name.  FTR, I'm not expecting you guys to do the renaming in this PR nor in this release; I understand it if you wouldl like to tackle this later.

> Ah yes I saw the `Type=dbus` in the docs, but a DBus API is still in the idea stage and would happen 6 months from now at the earliest. So I didn't think it was worth doing the work now to be a proper dbus service. Do you foresee problems of transitioning from type `notify` to `dbus` in the future?

Understood.  I just mentioned `Type=dbus` as a data point; also not expecting you guys to tackle this in this release :-).

Initially I don't foresee problems when you transition to `dbus` in the future, but I guess it will depend on a few details about the implementation.  For example, when using `Type=dbus` systemd will consider that the service is ready after it's acquired a name on the DBus bus, so that has to happen only when the service is really ready.  Another thing to take into consideration is the fact that systemd will send a `SIGTERM` to the service when the bus name is released.  I don't think these are problematic things; they're just implementation details.  But it's good to keep them in mind when designing the new API.  Worst case scenario, you can always stick to `Type=notify` if that's already working.